optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/DroolsConstraintSessionFactory.java

```diff
@@ -42,30 +49,65 @@
+    private final Map<String, org.drools.model.Rule> constraintToModelRuleMap;
 
-    public DroolsConstraintSessionFactory(SolutionDescriptor<Solution_> solutionDescriptor, KieBase kieBase,
+    public DroolsConstraintSessionFactory(SolutionDescriptor<Solution_> solutionDescriptor, Model model,
             List<DroolsConstraint<Solution_>> constraintList) {
         this.solutionDescriptor = solutionDescriptor;
-        this.kieBase = kieBase;
-        this.constraints = constraintList.stream()
-                .collect(toMap(constraint -> kieBase.getRule(constraint.getConstraintPackage(),
+        this.originalModel = model;
+        this.originalKieBase = KieBaseBuilder.createKieBaseFromModel(model);
+        this.activeKieBase = originalKieBase;
+        this.compiledRuleToConstraintMap = constraintList.stream()
+                .collect(toMap(constraint -> activeKieBase.getRule(constraint.getConstraintPackage(),
                         constraint.getConstraintName()), Function.identity()));
+        this.constraintToModelRuleMap = constraintList.stream()
+                .collect(toMap(Constraint::getConstraintId, constraint -> model.getRules().stream()
+                        .filter(rule -> Objects.equals(rule.getName(), constraint.getConstraintName()))
+                        .filter(rule -> Objects.equals(rule.getPackage(), constraint.getConstraintPackage()))
+                        .findFirst()
+                        .orElseThrow(() -> new IllegalStateException("Programming error: Rule for constraint (" +
+                                constraint + ") not found."))));
     }
 
     @Override
     public ConstraintSession<Solution_> buildSession(boolean constraintMatchEnabled, Solution_ workingSolution) {
         ScoreDefinition scoreDefinition = solutionDescriptor.getScoreDefinition();
         AbstractScoreHolder scoreHolder = (AbstractScoreHolder) scoreDefinition.buildScoreHolder(constraintMatchEnabled);
         scoreHolder.setJustificationListConverter((justificationList, rule) ->
-                matchJustificationsToOutput((List<Object>) justificationList, constraints.get(rule).getExpectedJustificationTypes()));
-        constraints.forEach((rule, constraint) -> scoreHolder.configureConstraintWeight(rule,
-                constraint.extractConstraintWeight(workingSolution)));
-        KieSession kieSession = kieBase.newKieSession();
-        ((RuleEventManager) kieSession).addEventListener(new OptaPlannerRuleEventListener()); // Enables undo in rules
+                matchJustificationsToOutput((List<Object>) justificationList,
+                        compiledRuleToConstraintMap.get(rule).getExpectedJustificationTypes()));
+        // Determine which rules to enable based on the fact that their constraints carry weight.
+        Score<?> zero = scoreDefinition.getZeroScore();
+        Set<String> enabledConstraintIdSet = new LinkedHashSet<>(compiledRuleToConstraintMap.size());
```

Comment:
I'll refactor this to `disabledConstraintIdSet` - that will cause less operations with the collection, and therefore perform better.